### ISSUE TITLE
Backups are uploaded, but the add-on doesn't seem to remember

### ISSUE THREAD

**Issue by cvissenb** [lavinir/hassio-onedrive-backup#253]
My backups are dutifully uploaded to OneDrive.
In the log, I can see the start and completion of the uploads.
My OneDrive shows the backup files and they look complete.
However, the UI shows all backups of the last 4 weeks to be local only.
And every day it re-uploads all of the "missing" backups to OneDrive.

**Comment by serl** [comment]
I think I have the same issue, here's the relevant logs:

```
11/23/2024 18:07:37 [15] Info: Found 1 backups to upload.
11/23/2024 18:07:37 [11] Info: Uploading hass_backup_2024-11-23-05-24 (11/23/2024 05:24:41)
11/23/2024 18:07:37 [11] Info: Fetching Local Backup (Slug:58b3692f)
11/23/2024 18:07:37 [11] Info: Backup (58b3692f) fetched successfully
11/23/2024 18:07:44 [11] Info: Starting file upload. (Size:277125120 bytes. Attempt: 1/3)
11/23/2024 18:09:24 [3] Info: Upload completed successfully
11/23/2024 18:12:37 [23] Info: Found 1 backups to upload.
11/23/2024 18:12:37 [23] Info: Uploading hass_backup_2024-11-23-05-24 (11/23/2024 05:24:41)
11/23/2024 18:12:37 [23] Info: Fetching Local Backup (Slug:58b3692f)
11/23/2024 18:12:37 [19] Info: Backup (58b3692f) fetched successfully
11/23/2024 18:12:43 [19] Info: Starting file upload. (Size:277125120 bytes. Attempt: 1/3)
11/23/2024 18:14:35 [23] Info: Upload completed successfully
11/23/2024 18:17:36 [31] Info: Found 1 backups to upload.
11/23/2024 18:17:36 [31] Info: Uploading hass_backup_2024-11-23-05-24 (11/23/2024 05:24:41)
11/23/2024 18:17:36 [31] Info: Fetching Local Backup (Slug:58b3692f)
11/23/2024 18:17:37 [31] Info: Backup (58b3692f) fetched successfully
11/23/2024 18:17:42 [30] Info: Starting file upload. (Size:277125120 bytes. Attempt: 1/3)
11/23/2024 18:19:14 [23] Info: Upload completed successfully
11/23/2024 18:22:36 [20] Info: Found 1 backups to upload.
11/23/2024 18:22:36 [20] Info: Uploading hass_backup_2024-11-23-05-24 (11/23/2024 05:24:41)
11/23/2024 18:22:36 [20] Info: Fetching Local Backup (Slug:58b3692f)
11/23/2024 18:22:36 [23] Info: Backup (58b3692f) fetched successfully
11/23/2024 18:22:42 [23] Info: Starting file upload. (Size:277125120 bytes. Attempt: 1/3)
```

**Comment by lavinir** [comment]
Could you please share the logs prior to upload and make sure they are set to verbose ?

**Comment by cvissenb** [comment]
Sure. I've set it to verbose now. Tomorrow I should have enough logging.

**Comment by cvissenb** [comment]
HI, thanks for looking into this.
I'm indeed running 2.3.1
But I see no indication of a new version. Possibly HASS lost the link to your repository.
I'll try to upgrade asap.

**Comment by cvissenb** [comment]
So far the end result is the same, just the logging looks different.
Is it safe to upload logs (because of lines like these)?
`Verbose: UploadSession Request: https://graph.microsoft.com/v1.0/drives/SOME-ID/items/SOME-ID%21105:/`
I also see a lot of this:
`Verbose: Attempting to read legacy backup description`

**Comment by cvissenb** [comment]
Here's the new log. I've replaced the IDs, just in case.
[onedrive2.zip](https://github.com/user-attachments/files/17925375/onedrive2.zip)

**Comment by cvissenb** [comment]
I did spot a suspicious error message:
`Error: Error saving backup metadata locally db0dfa40`
Does that explain the issue?

**Comment by lavinir** [comment]
@cvissenb  thanks.
The legacy backup metadata messages are because of the transition from saving the metadata in the OneDrive file description to locally. There were all kinds of issues with keeping it on OneDrive. So for backups that were created before 2.3.6 it will attempt to read the 'old' description.

Let's just try to compare that metadata for a backup that keeps being uploaded (hass_backup_2024-11-26-02-07;5cca3d64)
Could you please go to your OneDrive and look at the File Description for that backup and share it ?

Regarding the error saving metadata locally - I did not find that in the logs you attached.

**Comment by cvissenb** [comment]
Assuming you mean "Details", it's this:
`{"Slug":"5cca3d64","BackupDate":"2024-11-26T02:07:12.174487+01:00","InstanceName":null,"IsProtected":false,"Size":1876.68,"BackupType":"partial","Addons":[],"Folders":["addons/local","ssl"]}`

**Comment by lavinir** [comment]
Yes I meant details :)
Ok so just by looking at this it should recognize that the Backup Slug already exists and skip uploading. 

I can add some more debug logging here to try and understand better why that's not happening. Does this happen for newly created backups as well?

I know you have a lot of backups but do you mind maybe moving them all to some temporary folder to see if newly created backups are still doing this ?

**Comment by cvissenb** [comment]
Yes, it happens for every backup that is still on the HASS system, so everyday one more. (If still inside the keep window).
It does show the generational backups (youngest one is 212 days old) that are only online with the correct flag.

**Comment by lavinir** [comment]
And you're also seeing these backups in the addon dashboard right? Showing up as backed up on Onedrive as well ?

**Comment by cvissenb** [comment]
Yes, they have been backed up by the add-on. The naming is also correct.

I have moved the existing local backups to /backups/saved. Now HASS'es own backup list is empty.
I restarted the add-on.
However, the add-on still showed the locals backups to be uploaded. If I click the upload button on any, nothing happens. Not even an error message or something in the logs.
After that I triggered the backup. That finally removed the non-existing local backups from the UI.
Then I manually triggered the upload (normally it runs only at night).
Now waiting for the upload to complete.

**Comment by cvissenb** [comment]
And it has resumed re-uploading that last backup again.

**Comment by cvissenb** [comment]
Indeed, the OneDrive label is missing.
The content of `config/onlineBackupsData` is one file called `hass_backup_2024-11-27-14-45.aux`
```
{"Slug":"c8f31702","FileName":"hass_backup_2024-11-27-14-45.tar","BackupDate":"2024-11-27T14:45:17.928667+01:00","InstanceName":null,"Type":"partial","IsProtected":false,"Size":1820.13,"Addons":[],"Folders":["addons/local","ssl"],"TypeDisplayName":0}# 
```

And this is what it's currently doing:
```
11/28/2024 08:35:25 [18] Info: Found 1 backups to upload.
11/28/2024 08:35:25 [18] Verbose: Backup size to upload: 1.82GB
11/28/2024 08:35:25 [18] Info: Uploading hass_backup_2024-11-27-14-45 (11/27/2024 14:45:17)
11/28/2024 08:35:25 [18] Info: Fetching Local Backup (Slug:c8f31702)
11/28/2024 08:35:48 [24] Info: Backup (c8f31702) fetched successfully
11/28/2024 08:35:50 [24] Verbose: Backup Description: {"Slug":"c8f31702","BackupDate":"2024-11-27T14:45:17.928667+01:00","InstanceName":null,"IsProtected":false,"Size":1820.13,"BackupType":"partial","Addons":[],"Folders":["addons/local","ssl"]}
11/28/2024 08:35:51 [18] Verbose: UploadSession Request: https://graph.microsoft.com/v1.0/drives/SOME-ID/items/SOME-ID%21105:/hass_backup_2024-11-27-14-45.tar:/createUploadSession. Body: {}
11/28/2024 08:35:51 [18] Info: Starting file upload. (Size:1908541440 bytes. Attempt: 1/3)
11/28/2024 08:36:39 [24] Verbose: Uploaded 10%
11/28/2024 08:37:27 [24] Verbose: Uploaded 20%
11/28/2024 08:38:14 [10] Verbose: Uploaded 30%
11/28/2024 08:39:05 [24] Verbose: Uploaded 40%
```

**Comment by lavinir** [comment]
That's very weird.
Could you share the log from the start of the cycle - look for this line: "Refreshing existing backups..."

**Comment by cvissenb** [comment]
The log.
And again, thanks for looking into this.
[latest-1drive.zip](https://github.com/user-attachments/files/17944967/latest-1drive.zip)

**Comment by lavinir** [comment]
So you returned all the backups back? 
This is just going into a loop for the latest / new backups ?

I'm not sure why that's happening I will add some relevant debug logs to the preview channel and let you know when it's available.

**Comment by cvissenb** [comment]
No, the local backup are still in a sub-folder of /backup (I assume the add-on won't see them). There are a lot of old (generational) backups still on OneDrive.
So, currently there's just one local backup file.

**Comment by cvissenb** [comment]
Any new ideas?
BTW: I thought it was worth mentioning that I'm still on OS 11.4, this is because I had issues with the later releases.
(Like unexplained lagging and high CPU peaks.)
I don't know whether the OS version affects backups at all.

**Comment by lavinir** [comment]
Hey, sorry still the same idea , I need more logs just haven't managed to get that out to the preview ring yet. hopefully in the next couple of days.

**Comment by lavinir** [comment]
2.3.7 in the preview channel should have the additional logs. Please set to verbose logging and share again.

**Comment by cvissenb** [comment]
Thanks, I will try.
Can I switch to the preview channel while retaining settings and the OneDrive authentication/permission?

**Comment by cvissenb** [comment]
Installed the Preview channel and attached this morning's logs.
[onedrive_preview_log.zip](https://github.com/user-attachments/files/18057916/onedrive_preview_log.zip)

**Comment by lavinir** [comment]
Thanks @cvissenb . 
Very weird. Can you look in your Home Assistant configuration folder, there should be a "onlineBackupsData" subfolder inside.
Look for a file called: **hass_backup_2024-12-08-01-50.aux**
Could you see if the slug inside matches: **47bcf0ed** ?

**Comment by cvissenb** [comment]
Content of the onlineBackupsData folder:
```
drwxr-xr-x    2 root     root          4096 Dec  9 15:59 .
drwxr-xr-x   17 root     root          4096 Dec  7 18:15 ..
-rw-r--r--    1 root     root           250 Dec  9 03:52 hass_backup_2024-11-27-14-45.aux
-rw-r--r--    1 root     root           250 Dec  9 15:43 hass_backup_2024-11-29-01-05.aux
-rw-r--r--    1 root     root           250 Dec  9 15:35 hass_backup_2024-11-30-01-25.aux
-rw-r--r--    1 root     root           250 Dec  9 15:26 hass_backup_2024-12-01-01-55.aux
-rw-r--r--    1 root     root           250 Dec  9 15:17 hass_backup_2024-12-02-02-05.aux
-rw-r--r--    1 root     root           250 Dec  9 15:07 hass_backup_2024-12-03-02-45.aux
-rw-r--r--    1 root     root           250 Dec  9 14:59 hass_backup_2024-12-04-03-45.aux
-rw-r--r--    1 root     root           250 Dec  9 14:50 hass_backup_2024-12-05-04-10.aux
-rw-r--r--    1 root     root           249 Dec  9 14:41 hass_backup_2024-12-07-01-35.aux
-rw-r--r--    1 root     root           250 Dec  9 14:31 hass_backup_2024-12-08-01-50.aux
-rw-r--r--    1 root     root           258 Dec  9 15:52 hass_backup_2024-12-09-02-14.aux
```
Content of hass_backup_2024-12-09-02-14.aux:
```
{"Slug":"7259ce52","FileName":"hass_backup_2024-12-09-02-14.tar","BackupDate":"2024-12-09T02:14:24.684583+01:00","InstanceName":null,"Type":"partial","IsProtected":false,"Size":1741.92,"Addons":[],"Folders":["addons/local","media","ssl"],"TypeDisplayName":0}%
```

Content of hass_backup_2024-12-08-01-50.aux:
```
{"Slug":"47bcf0ed","FileName":"hass_backup_2024-12-08-01-50.tar","BackupDate":"2024-12-08T01:50:27.032813+01:00","InstanceName":null,"Type":"partial","IsProtected":false,"Size":1928.19,"Addons":[],"Folders":["addons/local","ssl"],"TypeDisplayName":0}
```

**Comment by lavinir** [comment]
Yes the aux file exists for that backup but it's still failing to register as an existing backup. I assume you don't see it in the dashboard right?

I need to add more logs to understand why it's not able to pick it up. (I'll publish another version to the preview channel today / tomorrow)

**Comment by cvissenb** [comment]
I see all local backups in the UI, but they miss the OneDrive label (and are continually uploaded).

**Comment by lavinir** [comment]
Yes that's what I meant.. missing OneDrive label means they are not registering as OneDrive backups. This is happening for all of your backups? None are showing up as on OneDrive ?

**Comment by cvissenb** [comment]
Indeed, except for some older ones (>200 days) that are **only** on OneDrive.

**Comment by serl** [comment]
Ah I think that's it. I have 203 and the first backup behaving strangely was the 201st...

**Comment by lavinir** [comment]
Yeah OK this is a bug. I checked the documentation, if more that 200 files are in a folder it will only return the first 200 and I'd need to send subsequent paging requests for the rest.

I'll implement the missing logic.

BTW, 200 backups.. Wow:)

**Comment by cvissenb** [comment]
I have quite a lot, only you cannot easily see the number of files in OneDrive's web-UI.
Could be because somehow I cannot get the generational backup working properly.
I'll keep them for now to verify whether your next release solves the issue :)
Thanks for looking into this.

**Comment by serl** [comment]
> I have quite a lot, only you cannot easily see the number of files in OneDrive's web-UI. 

It's easy when you know the trick:
- Scroll down and make sure you see the latest one
- Scroll back up and click the select all button (top left checkbox)
- Somewhere in the top right it says something like "203 selected"



Anyways yes thanks a lot for the addon and everything 🙂

**Comment by cvissenb** [comment]
Sorry, double post.

**Comment by cvissenb** [comment]
I've removed the surplus backups and 30 minutes later everything looks OK in the Web-UI.
No more endless uploading.
So, it was definitely the number of files on OneDrive.

Keep it open until the stable release that will fix this issue?

Anyway, thank you very much for diagnosing this.
Respect for your effort in creating this add-on.

**Comment by lavinir** [comment]
Great. I already have a fix for this ready for  the preview channel that will likely go out today in 2.3.8.

**Comment by sa0azs** [comment]
I still have the problem with files being uploaded but not shown as uploaded in UI.
I did a reinstallation of the addon, and after that it worked, once.
That file is uploaded, with the slug on the end of the filename.
The following are missing the slug, are uploaded but is shown av Local only in the UI.

I have like give (5) backups on OneDrive, so it is not the "too many files"-issue in my case.

Latest versions of everything relevant is installed.
